backend/store/etcd/environment_store_test.go

```diff
@@ -71,3 +67,38 @@ func TestEnvStorage(t *testing.T) {
 		assert.Equal(t, 1, len(envs))
 	})
 }
+
+func TestDeleteEnvironment_GH1567(t *testing.T) {
+	testWithEtcd(t, func(store store.Store) {
+		ctx := context.Background()
+		ctx = context.WithValue(ctx, types.OrganizationKey, "default")
+		ctx = context.WithValue(ctx, types.EnvironmentKey, "foo")
+
+		org := "default"
+		env := types.FixtureEnvironment("foo")
+		require.NoError(t, store.UpdateEnvironment(ctx, env))
+
+		asset := types.FixtureAsset("foo")
+		asset.Organization = org
+
+		require.NoError(t, store.UpdateAsset(ctx, asset))
+
+		// Assets do not apply to environments
+		require.NoError(t, store.DeleteEnvironment(ctx, env))
+	})
+}
+
+func TestDoubleDelete(t *testing.T) {
+	testWithEtcd(t, func(store store.Store) {
+		ctx := context.Background()
+		ctx = context.WithValue(ctx, types.OrganizationKey, "default")
+		ctx = context.WithValue(ctx, types.EnvironmentKey, "foo")
+
+		env := types.FixtureEnvironment("foo")
+		require.NoError(t, store.UpdateEnvironment(ctx, env))
+
+		// Second delete should succeed
+		require.NoError(t, store.DeleteEnvironment(ctx, env))
+		require.NoError(t, store.DeleteEnvironment(ctx, env))
```

Comment:
Could you help me understand why this is the desired action? If an environment has been deleted and does not exist, why wouldn't it return an error when you attempt to delete it again?

Reply:
It's because HTTP DELETE needs to be idempotent. See https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html for more information.